File: C++/Problems/Problems/problem_3805.cpp
#include "problem_3805.h"

#include <unordered_map>

bool problem_3805::test() {
    std::vector<std::string> words { "ab", "aa", "za", "aa" };

    int expected = 2;

    auto result = countPairs(words);

    std::cout << result << std::endl;

    return result == expected;
}

long long problem_3805::countPairs(std::vector<std::string> &words) {
    int m = words[0].size();
    long long result = 0;
    std::unordered_map<std::string, int> counts;
    for (auto w : words)
    {
        int b = w[0] - 'a';
        for (int i = 0; i < m; ++i)
            w[i] = (char)((w[i] - b + 26) % 26);
        if (counts.contains(w))
            result += counts[w]++;
        else
            counts[w] = 1;
    }
    return result;
}

File: Python/Problems/Problems/problem3805.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem


class Problem3805(Problem):
    def test(self) -> bool:
        words = ["ab", "aa", "za", "aa"]

        expected = 2

        result = self.countPairs(words)

        print(result)

        return result == expected

    def countPairs(self, words: List[str]) -> int:
        m = len(words[0])
        result = 0
        counts = dict()
        for w in words:
            a = list(w)
            b = ord(a[0]) - ord('a')
            for i in range(m):
                a[i] = chr((ord(a[i]) - b + 26) % 26)
            s = ''.join(a)
            if s in counts:
                result += counts[s]
                counts[s] += 1
            else:
                counts[s] = 1
        return result